### PR TITLE
feat(storage): implement functional downloadData

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

Initial implementations of upload data API with minimal unit test. This change includes some lower level change:
1. Remove the Axios compatible shapes and errors from XHR HTTP handler.
2. Remove use of `'events'` dependency to track the HTTP progress. Replace it with `onDownloadProgress` and `onUploadProgress` callbacks
3. Implement download task
4. Fix typing issues with `composeServiceApi` to fix issues of `useAccelerateEndpoint` config cannot be assigned.

#### Description of how you validated changes
Unit test

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/__tests__/providers/s3/downloadData.test.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+			getConfig: jest.fn(),
+		},
+	};
+});
+const credentials: Credentials = {
+	accessKeyId: 'accessKeyId',
+	sessionToken: 'sessionToken',
+	secretAccessKey: 'secretAccessKey',
+};
+const identityId = 'identityId';
```
Comment (kvramyasri7) [inline packages/storage/__tests__/providers/s3/downloadData.test.ts:30]: can we export this into different file as we have this to all apis?

File: packages/storage/__tests__/providers/s3/downloadData.test.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+jest.mock('../../../src/utils/transferTask');
+jest.mock('@aws-amplify/core', () => {
+	const core = jest.requireActual('@aws-amplify/core');
+	return {
+		...core,
+		AmplifyV6: {
+			...core.AmplifyV6,
+			Auth: {
+				fetchAuthSession: jest.fn(),
+			},
```
Comment (AllanZhengYP) [inline packages/storage/__tests__/providers/s3/downloadData.test.ts:20]: TODO: update this after PRs to import `fetchAuthSession` from `@aws-amplify/core` directly is merged.

File: packages/storage/__tests__/providers/s3/downloadData.test.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		jest.clearAllMocks();
+	});
+	const mockCreateDownloadTask = createDownloadTask as jest.Mock;
+	mockCreateDownloadTask.mockReturnValue('downloadTask');
+	(AmplifyV6.Auth.fetchAuthSession as jest.Mock).mockResolvedValue({
+		credentials,
+		identityId,
+	});
+	(AmplifyV6.getConfig as jest.Mock).mockReturnValue({
+		Storage: {
```
Comment (ashwinkumar6) [inline packages/storage/__tests__/providers/s3/downloadData.test.ts:44]: NIT:
I'm wondering if we should move all options, params, etc as constants above the describe method. Like how we have `credentials`.

Maybe these could all be generic configs defined on top. And then, according to the scenario
1.  downloadData from guest, we can override the generic config
2.  downloadData from private
3.  downloadData from protected

File: packages/storage/__tests__/providers/s3/downloadData.test.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+			{
+				credentials,
+				region: 'region',
+				useAccelerateEndpoint: true,
+				onDownloadProgress: onProgress,
+				abortSignal: expect.any(AbortSignal),
+			},
+			{
+				Bucket: 'bucket',
+				Key: 'protected/targetIdentityId/key',
```
Comment (ashwinkumar6) [inline packages/storage/__tests__/providers/s3/downloadData.test.ts:80]: NIT: 
same as above if we move the generic config on top, we could do something like
`${accessLevel}/${targetIdentityId}/${path}`

File: packages/storage/__tests__/providers/s3/downloadData.test.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+const credentials: Credentials = {
+	accessKeyId: 'accessKeyId',
+	sessionToken: 'sessionToken',
+	secretAccessKey: 'secretAccessKey',
+};
+const identityId = 'identityId';
+
+// TODO: test validation errors
+describe('downloadData', () => {
+	beforeEach(() => {
```
Comment (ashwinkumar6) [inline packages/storage/__tests__/providers/s3/downloadData.test.ts:34]: NIT:
Would it be good to mock functions on `beforeEach` and clear mocks on `afterEach`

Example
```
	beforeEach(() => {
		(fetchAuthSession as jest.Mock).mockResolvedValue({
			credentials,
			identityId: targetIdentityId,
		});
		(AmplifyV6.getConfig as jest.Mock).mockReturnValue({
			Storage: {
				bucket,
				region,
			},
		});
	});
	afterEach(() => {
		jest.clearAllMocks();
	});
```

**Review by kvramyasri7 — DISMISSED** [review]
Looks good to me. Please extract the common mocks and some more testing 👍

**Review by kvramyasri7 — COMMENTED** [review]
Some merge conflicts on the PR. Please take are of them and everything looks good

File: packages/storage/src/providers/s3/apis/downloadData.ts [hunk 6]
```diff
@@ -3,8 +3,102 @@
...
+				useAccelerateEndpoint,
+			},
+			{
+				Bucket: bucket,
+				Key: finalKey,
+			}
+		);
+		validateS3RequiredParameter(!!body, 'Body');
+		return {
+			body,
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/apis/downloadData.ts:96]: nit: is `ContentLength` missed in result cause we are extracting it in the result?

**Review by kvramyasri7 — APPROVED** [review]
Looks good to me,  Allan🏅.

**Review by ashwinkumar6 — DISMISSED** [review]

**Review by ashwinkumar6 — APPROVED** [review]
LGTM

**Review by kvramyasri7 — APPROVED** [review]

**Pull request merged**